src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+u16 cs8900a_device::cs8900_read_register(u16 ppaddress)
+{
+  u16 value = GET_PP_16(ppaddress);
+  
+  /* --- check the register address --- */
+  if(ppaddress<0x100) {
+    /* reserved range reads 0x0300 on real HW */
+    if((ppaddress>=0x0004)&&(ppaddress<0x0020)) {
+      return 0x0300;
+    }
+  }
+  
+  /* --- read control register range --- */
+  else if(ppaddress<0x120) {
+    u16 regNum = ppaddress - 0x100;
+    regNum &= ~1;
+    regNum ++;
+#ifdef CS8900_DEBUG_REGISTERS
+    log_message(cs8900_log,
+                "Read  Control Register %04x: %04x (reg=%02x)",
+                ppaddress,value,regNum);
+#endif
+ 
+    /* reserved register? */
+    if((regNum==0x01)||(regNum==0x11)||(regNum>0x19)) {
+#ifdef CS8900_DEBUG_WARN_REG
+      log_message(cs8900_log,
+                  "WARNING! Read reserved Control Register %04x (reg=%02x)",
+                  ppaddress,regNum);
+#endif
+      /* real HW returns 0x0300 in reserved register range */
+      return 0x0300;
+    }
+
+    /* make sure interal address is always valid */
+    assert((value&0x3f) == regNum); 
+  }
+  
+  /* --- read status register range --- */
+  else if(ppaddress<0x140) {
+    u16 regNum = ppaddress - 0x120;
+    regNum &= ~1;
+#ifdef CS8900_DEBUG_REGISTERS
+#ifdef CS8900_DEBUG_IGNORE_RXEVENT
+    if(regNum!=4) // do not show RXEVENT
+#endif
+    log_message(cs8900_log,
+                "Read  Status  Register %04x: %04x (reg=%02x)",
+                ppaddress,value,regNum);
+#endif
+
+    /* reserved register? */
+    if((regNum==0x02)||(regNum==0x06)||(regNum==0x0a)||
+       (regNum==0x0e)||(regNum==0x1a)||(regNum==0x1e)) {
+#ifdef CS8900_DEBUG_WARN_REG
+      log_message(cs8900_log,
+                  "WARNING! Read reserved Status Register %04x (reg=%02x)",
+                  ppaddress,regNum);
+#endif
+      /* real HW returns 0x0300 in reserved register range */
+      return 0x0300;
+    }
+
+    /* make sure interal address is always valid */
+    assert((value&0x3f) == regNum);
+  }
+  
+  /* --- read transmit register range --- */
+  else if(ppaddress<0x150) {
+    if(ppaddress==0x144) {
+      /* make sure interal address is always valid */
+      assert((value&0x3f) == 0x09);
+#ifdef CS8900_DEBUG_REGISTERS
+      log_message(cs8900_log,
+                  "Read  TX Cmd  Register %04x: %04x",
+                  ppaddress,value);
+#endif
+    }
+    else if(ppaddress==0x146) {
+#ifdef CS8900_DEBUG_REGISTERS
+      log_message(cs8900_log,
+                  "Read  TX Len  Register %04x: %04x",
+                  ppaddress,value);
+#endif
+    }
+    /* reserved range */
+    else {
+#ifdef CS8900_DEBUG_WARN_REG
+      log_message(cs8900_log,
+                  "WARNING! Read reserved Initiate Transmit Register %04x",
+                  ppaddress);
+#endif
+      /* real HW returns 0x0300 in reserved register range */
+      return 0x0300;
+    }
+  }
+  
+  /* --- read address filter register range --- */
+  else if(ppaddress<0x160) {
```

Comment:
Please don’t put extra lines between the closing `}` of a compound statement controlled by an `if` and an `else` – it can easily lead to confusion.  If a developer has this at the bottom of their window/terminal they will see the blank line and assume that there’s no following `else`.